### PR TITLE
Zombie Mode [New Game Mode]

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR 
Adds the Zombie gamemode, a new gamemode with new antags. It begins with a few infected players at the beginning before spiraling out to a full blown outbreak. It also features several round-ending events to try and contain the outbreak.

Don't Dead, ~shitcode~ Open Inside

## **Screenshots**
image
image

## **Changelog**

:cl:
- add: Added the Zombie game mode as a low chance option in Secret.
- add: Added the CBURN Unit, an advanced operations team for Chemical, Biological, Undead, Radiological, and Nuclear threats.
- tweak: Non-human zombies no longer get a damage buff.
- tweak: Zombie infection chance is now based on the amount of clothing worn.
- tweak: Zombies have had a visual overhaul.
- fix: Zombies now do less damage to each other rather than more.

### CODE REVIEW

**Comment by moonheart08** [comment]
Should we roll event zombies out of rotation in favor of this mode? Or, more interestingly, should event zombies start this game rule.

**Comment by EmoGarbage404** [comment]
> Should we roll event zombies out of rotation in favor of this mode? Or, more interestingly, should event zombies start this game rule.

I didn't know that you could activate game rules mid-round like that but holy fuck yes yes do that what the hell that's amazing. The only thing i'd have to comment on is that the event time frame would have to be decreased from 50+ minutes in order to account for the incubation period of the new virus.

**Comment by moonheart08** [comment]
> 

You'll have to code the game rule to support being activated midround, but yes that's a thing.

File: Resources/Prototypes/secret_weights.yml [hunk 1]
```diff
@@ -3,4 +3,5 @@
   weights:
     Extended: 0.25
     Nukeops: 0.25
-    Traitor: 0.75
\ No newline at end of file
+    Traitor: 0.75
+    Zombie: 0.10
```
Comment (moonheart08) [inline Resources/Prototypes/secret_weights.yml:7]: Should probably be about 0.05, to drop it's chances of occuring to about once-twice a day.

File: Content.Server/GameTicking/Rules/ZombieRuleSystem.cs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    [Dependency] private readonly IChatManager _chatManager = default!;
+    [Dependency] private readonly IPlayerManager _playerManager = default!;
+    [Dependency] private readonly IServerPreferencesManager _prefs = default!;
+    [Dependency] private readonly RoundEndSystem _roundEndSystem = default!;
+    [Dependency] private readonly DiseaseSystem _diseaseSystem = default!;
+    [Dependency] private readonly PopupSystem _popup = default!;
+    [Dependency] private readonly ActionsSystem _action = default!;
+    [Dependency] private readonly ZombifyOnDeathSystem _zombify = default!;
+
+    private Dictionary<string, string> _initialInfected = new();
```
Comment (mirrorcult) [inline Content.Server/GameTicking/Rules/ZombieRuleSystem.cs:45]: make this clearer that it's storing their names

File: Content.Server/GameTicking/Rules/ZombieRuleSystem.cs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+                totalPlayers.Add(ent.Item2.Owner);
+
+                if (HasComp<ZombieComponent>(ent.Item1.Owner))
+                    livingZombies.Add(ent.Item2.Owner);
+                else
+                    livingHumans.Add(ent.Item2.Owner);
+            }
+        }
+        return (FixedPoint2) livingZombies.Count / (FixedPoint2) totalPlayers.Count;
+    }
```
Comment (mirrorcult) [inline Content.Server/GameTicking/Rules/ZombieRuleSystem.cs:216]: why is the percentage fixedpoint2 and not just float? it wasn't really intended for this. if its to round, then you should just use float and round for displaying. might also need to use `MathHelper.CloseTo` for the percentage checks if it doesn't work cause of float nonsense

File: Content.Server/Zombies/ZombieSystem.cs [hunk 4]
```diff
@@ -15,13 +19,42 @@ public sealed class ZombieSystem : EntitySystem
...
+
+            if (!TryComp<InventoryComponent>(uid, out var inventoryComponent))
+                return baseChance;
+
+            if (!_inv.TryGetContainerSlotEnumerator(uid, out var enumerator, inventoryComponent))
+                return baseChance;
+
+            var items = 0f;
+            var total = 0f;
+            while  (enumerator.MoveNext(out var con))
```
Comment (mirrorcult) [inline Content.Server/Zombies/ZombieSystem.cs:44]:
```suggestion
            while (enumerator.MoveNext(out var con))
```

File: Content.Shared/CCVar/CCVars.cs [hunk 5]
```diff
@@ -286,6 +286,19 @@ public sealed class CCVars : CVars
...
 
+        /*
+         * Zombie
+         */
+
+        public static readonly CVarDef<int> ZombieMinPlayers =
+            CVarDef.Create("zombie.min_players", 25);
+
+        public static readonly CVarDef<int> ZombieMaxInfected =
+            CVarDef.Create("zombie.max_infected", 4);
```
Comment (mirrorcult) [inline Content.Shared/CCVar/CCVars.cs:297]: max initial infected i assume?

File: Content.Shared/CCVar/CCVars.cs [hunk 6]
```diff
@@ -286,6 +286,19 @@ public sealed class CCVars : CVars
...
+
+        public static readonly CVarDef<int> ZombiePlayersPerInfected =
+            CVarDef.Create("zombie.players_per_infected", 7);
```
Comment (mirrorcult) [inline Content.Shared/CCVar/CCVars.cs:300]: isnt it basically always gonna be 4 then, since the min is 25 and it would max out at 28

File: Content.Shared/Zombies/ZombieEvents.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
+using Content.Shared.Actions;
+
+namespace Content.Shared.Zombies;
+
+/// <summary>
+///     Event raised whenever an entity is zombified.
+///     Used by the zombie gamemode to track total infections.
+/// </summary>
```
Comment (mirrorcult) [inline Content.Shared/Zombies/ZombieEvents.cs:8]: raised directed or broadcast or both?

File: Resources/Locale/en-US/game-ticking/game-presets/preset-zombies.ftl [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+zombie-end-nuke-armed-event-fail = The nuke was not detonated.
+zombie-end-nuke-armed-event-success = The nuke was successfully detonated by {$name} ({$username}), stopping the outbreak.
+
+zombie-alone = You feel entirely alone.
+
+zombie-round-end-initial-count = {$initialCount ->
+    [one] There was one initial infected:
+    *[other] There were {$initialCount} initial infected:
+}
+zombie-round-end-user-was-initial = - [color=Plum]{$name}[/color] ([color=gray]{$username}[/color]) was one of the initial infected.
```
Comment (mirrorcult) [inline Resources/Locale/en-US/game-ticking/game-presets/preset-zombies.ftl:19]:
```suggestion
zombie-round-end-user-was-initial = - [color=plum]{$name}[/color] ([color=gray]{$username}[/color]) was one of the initial infected.
```

idk if being capitalized actually breaks it but uhhh

File: Resources/Locale/en-US/zombies/zombie.ftl [hunk 9]
```diff
@@ -2,6 +2,6 @@ zombie-transform = {CAPITALIZE(THE($target))} turned into a zombie!
 zombie-infection-greeting = You have become a zombie. Your goal is to seek out the living and to try to infect them. Work together with your fellow zombies to overpower the remaining crewmates.
 
 zombie-generic = zombie
-zombie-name-prefix = zombified {$target}
+zombie-name-prefix = Zombie {$target}
```
Comment (mirrorcult) [inline Resources/Locale/en-US/zombies/zombie.ftl:5]: why the change? i liked "zombified"

File: Resources/Locale/en-US/game-ticking/game-presets/preset-zombies.ftl [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+zombie-description = A virus able to animate the dead has been unleashed unto the station! Work with your crewmates to contain the outbreak and survive.
+
+zombie-not-enough-ready-players = Not enough players readied up for the game! There were {$readyPlayersCount} players readied up out of {$minimumPlayers} needed. Can't start Zombies.
+zombie-no-one-ready = No players readied up! Can't start Zombies.
+
+zombie-patientzero-role-greeting = You are patient 0. Hide your infection, get supplies, and be prepared to turn once you die.
+
+zombie-nuke-armed-event = In response to the zombie outbreak, we have armed the nuclear device positioned onboard the station. A disk has been automatically teleported inside of it. The code to detonate the device is: {$code}.
+zombie-end-nuke-armed-event-fail = The nuke was not detonated.
+zombie-end-nuke-armed-event-success = The nuke was successfully detonated by {$name} ({$username}), stopping the outbreak.
```
Comment (mirrorcult) [inline Resources/Locale/en-US/game-ticking/game-presets/preset-zombies.ftl:11]: i dont see any actual code for this? nor for spawning CBURN stuff

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Content.Server/GameTicking/Rules/ZombieRuleSystem.cs [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+    {
+        _zombify.ZombifyEntity(uid);
+
+        var action = new InstantAction(_prototypeManager.Index<InstantActionPrototype>(ZombifySelfActionPrototype));
+        _action.RemoveAction(uid, action);
+    }
+
+    private float GetInfectedPercentage(out List<EntityUid> livingHumans)
+    {
+        var allPlayers = EntityQuery<HumanoidAppearanceComponent, MobStateComponent>();
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/ZombieRuleSystem.cs:198]:
```suggestion
        var allPlayers = EntityQuery<HumanoidAppearanceComponent, MobStateComponent>(true);
```

File: Content.Server/GameTicking/Rules/ZombieRuleSystem.cs [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+        var totalPlayers = new List<EntityUid>();
+        var livingZombies = new List<EntityUid>();
+        livingHumans = new();
+        foreach (var ent in allPlayers)
+        {
+            if (ent.Item2.IsAlive())
+            {
+                totalPlayers.Add(ent.Item2.Owner);
+
+                if (HasComp<ZombieComponent>(ent.Item1.Owner))
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/ZombieRuleSystem.cs:208]: Should probably use an entityquery for zombiecomponent and check HasComp with that.

File: Content.Server/GameTicking/Rules/ZombieRuleSystem.cs [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+        {
+            if (player.AttachedEntity != null)
+            {
+                playerList.Add(player);
+
+                var pref = (HumanoidCharacterProfile) _prefs.GetPreferences(player.UserId).SelectedCharacter;
+                if (pref.AntagPreferences.Contains(PatientZeroPrototypeID))
+                    prefList.Add(player);
+            }
+        }
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/ZombieRuleSystem.cs:241]: Would this try to infect people in lobby?

File: Content.Server/Zombies/ZombieSystem.cs [hunk 14]
```diff
@@ -15,13 +19,42 @@ public sealed class ZombieSystem : EntitySystem
...
+            var items = 0f;
+            var total = 0f;
+            while (enumerator.MoveNext(out var con))
+            {
+                total++;
+                if (con.ContainedEntity != null)
+                    items++;
+            }
+
+            var max = component.MaxZombieInfectionChance;
```
Comment (metalgearsloth) [inline Content.Server/Zombies/ZombieSystem.cs:51]: This is pretty weird because stuff like pocket items will block infection chances.

File: Content.Server/Zombies/ZombifyOnDeathSystem.cs [hunk 15]
```diff
@@ -53,95 +60,135 @@ public override void Initialize()
...
+        /// </remarks>
         public void ZombifyEntity(EntityUid target)
         {
+            //Don't zombfiy zombies
             if (HasComp<ZombieComponent>(target))
                 return;
 
-            _disease.CureAllDiseases(target);
+            //you're a real zombie now, son.
+            var zombiecomp = EnsureComp<ZombieComponent>(target);
```
Comment (metalgearsloth) [inline Content.Server/Zombies/ZombifyOnDeathSystem.cs:89]: Just AddComp it as you know they don't have.

File: Content.Shared/Zombies/ZombieEvents.cs [hunk 16]
```diff
@@ -0,0 +1,25 @@
+using Content.Shared.Actions;
+
+namespace Content.Shared.Zombies;
+
+/// <summary>
+///     Event that is broadcast whenever an entity is zombified.
+///     Used by the zombie gamemode to track total infections.
+/// </summary>
+public sealed class EntityZombifiedEvent : EventArgs
```
Comment (metalgearsloth) [inline Content.Shared/Zombies/ZombieEvents.cs:9]:
```suggestion
public sealed class EntityZombifiedEvent : EntityEventArgs
```

Or if you want to be based make it a struct ref event.

**Review by metalgearsloth — APPROVED** [review]

File: Content.Server/GameTicking/Rules/ZombieRuleSystem.cs (hunk 13 above)
Comment (EmoGarbage404) [reply Content.Server/GameTicking/Rules/ZombieRuleSystem.cs:241]: No, because the AttachedEntity != check fails, removing them from the playerList pool. I tested this multiple times locally and couldn't get it to happen once.

File: Content.Server/Zombies/ZombieSystem.cs (hunk 14 above)
Comment (EmoGarbage404) [reply Content.Server/Zombies/ZombieSystem.cs:51]: I'm aware that this is kind of weird, but it doesn't really impact much. It can also be considered Depth:tm:

**Comment by Just-a-Unity-Dev** [comment]
something I would recommend is showing a popup rather than a chat message

**Comment by EmoGarbage404** [comment]
> something I would recommend is showing a popup rather than a chat message

there's a popup already when you turn into a zombie. It's been like that for a while.

**Comment by Just-a-Unity-Dev** [comment]
> > something I would recommend is showing a popup rather than a chat message
> 
> there's a popup already when you turn into a zombie. It's been like that for a while.

ah i must be a dumbass then lmao, lgtm

**Comment by EmoGarbage404** [comment]
i am free.
